experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+        val >>= 1;
+      }
+      llvm::outs() << " ";
+    }
+
+    llvm::outs() << ins->valid << " " << ins->ready << "\n";
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Support
+//===----------------------------------------------------------------------===//
+/// Components required for the internal state.
+
+class ForkSupport {
+public:
+  ForkSupport(unsigned size, unsigned datawidth = 0)
+      : size(size), datawidth(datawidth) {
+    transmitValue.resize(size, false);
+    keepValue.resize(size, false);
+    blockStopArray.resize(size, false);
+  }
+
+  void resetDataless(ConsumerRW *ins, std::vector<ProducerRW *> &outs) {
+    for (unsigned i = 0; i < size; ++i) {
+      transmitValue[i] = true;
+      keepValue[i] = !outs[i]->ready;
+      outs[i]->valid = ins->valid;
+      blockStopArray[i] = !outs[i]->ready;
+    }
+
+    // or_n
+    anyBlockStop = false;
+    for (bool c : blockStopArray)
+      anyBlockStop = anyBlockStop || c;
+
+    ins->ready = !anyBlockStop;
+    backpressure = ins->valid && anyBlockStop;
+  }
+
+  void execDataless(bool isClkRisingEdge, ConsumerRW *ins,
+                    std::vector<ProducerRW *> &outs) {
+    for (unsigned i = 0; i < outs.size(); ++i) {
+      keepValue[i] = !outs[i]->ready && transmitValue[i];
+      if (isClkRisingEdge)
+        transmitValue[i] = keepValue[i] || !backpressure;
+      outs[i]->valid = transmitValue[i] && ins->valid;
+      blockStopArray[i] = keepValue[i];
+    }
+
+    // or_n
+    anyBlockStop = false;
+    for (bool c : blockStopArray)
+      anyBlockStop = anyBlockStop || c;
+
+    ins->ready = !anyBlockStop;
+    backpressure = ins->valid && anyBlockStop;
+  }
+
+  void reset(ConsumerRW *ins, std::vector<ProducerRW *> &outs,
+             const ConsumerData &insData, std::vector<ProducerData> &outsData) {
+    resetDataless(ins, outs);
+    for (auto &out : outsData)
+      out = insData;
+  }
+
+  void exec(bool isClkRisingEdge, ConsumerRW *ins,
+            std::vector<ProducerRW *> &outs, const ConsumerData &insData,
+            std::vector<ProducerData> &outsData) {
+    execDataless(isClkRisingEdge, ins, outs);
+    for (auto &out : outsData)
+      out = insData;
+  }
+
+private:
+  unsigned size;
+  // eager fork
+  std::vector<bool> transmitValue, keepValue;
+  // fork dataless
+  std::vector<bool> blockStopArray;
+  bool anyBlockStop = false, backpressure = false;
+
+public:
+  unsigned datawidth;
+};
+
+class JoinSupport {
+public:
+  JoinSupport(unsigned size) : size(size) {}
+
+  void exec(std::vector<ConsumerRW *> &ins, ProducerRW *outs) {
+    outs->valid = true;
+    for (unsigned i = 0; i < size; ++i)
+      outs->valid = outs->valid && ins[i]->valid;
+
+    for (unsigned i = 0; i < size; ++i) {
+      ins[i]->ready = outs->ready;
+      for (unsigned j = 0; j < size; ++j)
+        if (i != j)
+          ins[i]->ready = ins[i]->ready && ins[j]->valid;
+    }
+  }
+
+private:
+  unsigned size;
+};
```

Comment:
This is very good. That's the actual exec model for a join yet it is separate from an actual execution model, so it can be used independently and the actual join execution model (`JoinModel`) is now trivial. Good job!